### ISSUE TITLE
Incorrect quark function for GOptionError

### ISSUE THREAD

**Issue by tmiasko** [gtk-rs/gir#523]
Gir selects incorrect quark function for GOptionError from glib. It decides to use `g_option_context_error_quark` whereas the correct one is `g_option_error_quark`.  

It happens because gir XML for GOptionError contains no associated quark function, and so generator falls back to error domain, which in turn contains quite unusual value of `g-option-context-error-quark`.

**Comment by sdroege** [comment]
This should probably also be fixed in the GLib-2.0.gir. Can you report it there too: https://bugzilla.gnome.org/enter_bug.cgi?product=glib

**Comment by tmiasko** [comment]
I did a little bit of detective work before reporting this upstream.  It seems
that this is working as designed. The glib:error-domain attribute is not
supposed to hold a function returning quark corresponding to given error
domain, but rather it should hold the string itself. This was changed in 2011,
in the following commit:

> Switch to storing string form of error quarks
>     
> Instead of storing the name of the function to call to get the
> error quark, store the string form of the error quark, which
> we derive from the introspection binary during scanning.
> 
> Update EnumBlob and GIEnumInfo to include the new information.
> 
> This will allow determining a back-mapping from error quark
> to error domain without having to dlsym() and call all the
> known error quark functions.
> 
> Based on earlier patches from Owen Taylor and Maxim Ermilov.
> 
> https://bugzilla.gnome.org/show_bug.cgi?id=602516

I suppose we could always try to make our own quark with g_quark_from_string
with some static caching in similar way it is done by G_DEFINE_QUARK. Though it
is quite unfortunate.

**Comment by EPashkin** [comment]
@tmiasko Thanks for investigation.
Agree with possible solution.
There also cases when enum contains direct function definition "quark" (Ex. GtkBuilderError), currently `src/analysis/enums.rs:get_error_quark_name` prefer this function over "error_domain" attribute.
Seems we need/can works with 2 different types of domains getting 😭

**Comment by tmiasko** [comment]
IMHO the best solution would be to have quark functions always available in gir
xml for error enums. This is somewhat wishful thinking on my part, but I opened
following report with a patch: https://bugzilla.gnome.org/show_bug.cgi?id=793219

BTW, even if there is quark function we don't have to use it, so it is not like
we have to implement both versions to make it work.

**Comment by sdroege** [comment]
Yeah, we will only have to use the actual quark name, always. But that would require everything to depend on lazy_static ideally, calling g_quark_from_string() or g_quark_from_static_string() is not super fast and always goes through a global mutex.
